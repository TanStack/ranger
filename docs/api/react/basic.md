---
name: Basic
route: /api/basic
menu: API
---

## Examples
Want to skip to the implementation? Check out these examples:

- [basic](../../examples/react/basic)

The API below described how to use the **basic** features.

## Options

### values

```tsx
values: ReadonlyArray<number>
```
**Required** The current value (or values) for the range.

### min

```tsx
min: number
```
**Required** The minimum limit for the range.

### max

```tsx
max: number
```
**Required** The maximum limit for the range.

### stepSize

```ts
stepSize: number
```
**Required** The distance between selectable steps.

### onChange

```ts
onChange: (instance: Ranger<TTrackElement>) => void
```
A function that is called when the handle is released.

## API

### handles
```tsx
handles: ReadonlyArray<{value: number; isActive: boolean; onKeyDownHandler(event): function; onMouseDownHandler(event): function; onTouchStart(event): function}>
```
Handles to be rendered. Each `handle` has the following props:
 - `value: number` - The current value for the handle.
 - `isActive: boolean` - Denotes if the handle is currently being dragged.
 - `onKeyDownHandler(event): func`
 - `onMouseDownHandler(event): func`
 - `onTouchStart(event): func`

### activeHandleIndex
```tsx
activeHandleIndex: null | number
```
The zero-based index of the handle that is currently being dragged, or `null` if no handle is being dragged.
