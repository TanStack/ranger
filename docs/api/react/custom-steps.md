---
name: Custom Steps
route: /api/custom-steps
menu: API
---

## Examples
Want to skip to the implementation? Check out these examples:

- [custom-steps](../examples/custom-steps)

The API below described how to use the **custom-steps** features.

## Options

### steps

```ts
steps: Array<number>
```
An array of custom steps to use. This will override `stepSize`

### ticks

```ts
ticks: Array<number>
```
An array of custom ticks to use. This will override `tickSize`

## API

### getTicks
```tsx
getTicks: () => ReadonlyArray<{value: number; key: number; percentage: number}>
```
Ticks to be rendered. Each `tick` has the following props:
  - `value: number` - The tick number to be displayed
  - `key: number` - The key of a tick
  - `percentage: number` - Percentage value of where tick should be placed on ranger
