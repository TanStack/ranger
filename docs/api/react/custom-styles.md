---
name: Custom Styles
route: /api/custom-styles
menu: API
---

## Examples
Want to skip to the implementation? Check out these examples:

- [custom-styles](../examples/custom-styles)

The API below described how to use the **custom-steps** features.

## API

### getSegments
```tsx
getSegments: () => ReadonlyArray<{left: number; width: number}>
```
Segments to be rendered. Each `segment` has the following props:
  - `left: number` - Percentage value of where segment should start on ranger
  - `width: number` - Percentage value of segment width 
