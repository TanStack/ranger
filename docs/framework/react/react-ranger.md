---
title: React Ranger
---

You can install TanStack Ranger with any [NPM](https://npmjs.com) package manager.

```sh
npm install @tanstack/react-ranger
```
