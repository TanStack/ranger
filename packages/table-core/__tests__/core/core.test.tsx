import * as React from 'react'

// import { renderHook } from '@testing-library/react-hooks'
import * as RTL from '@testing-library/react'
import { createTable, useTable } from '@tanstack/react-table'
import { act, renderHook } from '@testing-library/react-hooks'

type Row = {
  firstName: string
  lastName: string
  age: number
  visits: number
  status: string
  progress: number
}

const table = createTable<{ Row: Row }>()

const defaultData: Row[] = [
  {
    firstName: 'tanner',
    lastName: 'linsley',
    age: 29,
    visits: 100,
    status: 'In Relationship',
    progress: 50,
  },
  {
    firstName: 'derek',
    lastName: 'perkins',
    age: 40,
    visits: 40,
    status: 'Single',
    progress: 80,
  },
  {
    firstName: 'joe',
    lastName: 'bergevin',
    age: 45,
    visits: 20,
    status: 'Complicated',
    progress: 10,
  },
]

const defaultColumns = table.createColumns([
  table.createGroup({
    header: 'Name',
    footer: props => props.column.id,
    columns: [
      table.createDataColumn('firstName', {
        cell: info => info.value,
        footer: props => props.column.id,
      }),
      table.createDataColumn(row => row.lastName, {
        id: 'lastName',
        cell: info => info.value,
        header: () => <span>Last Name</span>,
        footer: props => props.column.id,
      }),
    ],
  }),
  table.createGroup({
    header: 'Info',
    footer: props => props.column.id,
    columns: [
      table.createDataColumn('age', {
        header: () => 'Age',
        footer: props => props.column.id,
      }),
      table.createGroup({
        header: 'More Info',
        columns: [
          table.createDataColumn('visits', {
            header: () => <span>Visits</span>,
            footer: props => props.column.id,
          }),
          table.createDataColumn('status', {
            header: 'Status',
            footer: props => props.column.id,
          }),
          table.createDataColumn('progress', {
            header: 'Profile Progress',
            footer: props => props.column.id,
          }),
        ],
      }),
    ],
  }),
])

describe('core', () => {
  it('renders a table with markup', () => {
    const Table = () => {
      const [data] = React.useState<Row[]>(() => [...defaultData])
      const [columns] = React.useState<typeof defaultColumns>(() => [
        ...defaultColumns,
      ])
      const [columnVisibility, setColumnVisibility] = React.useState({})

      const rerender = React.useReducer(() => ({}), {})[1]

      const instance = useTable(table, {
        data,
        columns,
        onColumnVisibilityChange: setColumnVisibility,
        state: {
          columnVisibility,
        },
        // debug: true,
      })

      return (
        <div className="p-2">
          <table {...instance.getTableProps()}>
            <thead>
              {instance.getHeaderGroups().map(headerGroup => (
                <tr {...headerGroup.getHeaderGroupProps()}>
                  {headerGroup.headers.map(header => (
                    <th {...header.getHeaderProps()}>
                      {header.isPlaceholder ? null : header.renderHeader()}
                    </th>
                  ))}
                </tr>
              ))}
            </thead>
            <tbody {...instance.getTableBodyProps()}>
              {instance.getRowModel().rows.map(row => (
                <tr {...row.getRowProps()}>
                  {row.getVisibleCells().map(cell => (
                    <td {...cell.getCellProps()}>{cell.renderCell()}</td>
                  ))}
                </tr>
              ))}
            </tbody>
            <tfoot>
              {instance.getFooterGroups().map(footerGroup => (
                <tr {...footerGroup.getFooterGroupProps()}>
                  {footerGroup.headers.map(header => (
                    <th {...header.getFooterProps()}>
                      {header.isPlaceholder ? null : header.renderFooter()}
                    </th>
                  ))}
                </tr>
              ))}
            </tfoot>
          </table>
          <div className="h-4" />
          <button onClick={() => rerender()} className="border p-2">
            Rerender
          </button>
        </div>
      )
    }

    // let api: any;

    const rendered = RTL.render(<Table />)

    // RTL.screen.logTestingPlaygroundURL();

    RTL.screen.getByRole('table')
    expect(RTL.screen.getAllByRole('rowgroup').length).toEqual(3)
    expect(RTL.screen.getAllByRole('row').length).toEqual(9)
    expect(RTL.screen.getAllByRole('columnheader').length).toEqual(12)
    expect(RTL.screen.getAllByRole('columnfooter').length).toEqual(12)
    expect(RTL.screen.getAllByRole('gridcell').length).toEqual(18)

    expect(
      Array.from(rendered.container.querySelectorAll('thead > tr')).map(d =>
        Array.from(d.querySelectorAll('th')).map(d => [
          d.innerHTML,
          d.getAttribute('colspan'),
        ])
      )
    ).toMatchSnapshot()

    expect(
      Array.from(rendered.container.querySelectorAll('tbody > tr')).map(d =>
        Array.from(d.querySelectorAll('td')).map(d => d.innerHTML)
      )
    ).toMatchSnapshot()

    expect(
      Array.from(rendered.container.querySelectorAll('tfoot > tr')).map(d =>
        Array.from(d.querySelectorAll('th')).map(d => [
          d.innerHTML,
          d.getAttribute('colspan'),
        ])
      )
    ).toMatchSnapshot()
  })

  it('has a stable api', () => {
    const { result } = renderHook(() => {
      const rerender = React.useReducer(() => ({}), {})[1]

      const instance = useTable(table, {
        data: defaultData,
        columns: defaultColumns,
      })

      return {
        instance,
        rerender,
      }
    })

    const prev = result.current

    act(() => {
      result.current.rerender()
    })

    const next = result.current

    expect(prev).toStrictEqual(next)
  })

  it('can return the rowModel', () => {
    const { result } = renderHook(() => {
      const rerender = React.useReducer(() => ({}), {})[1]

      const instance = useTable(table, {
        data: defaultData,
        columns: defaultColumns,
      })

      return {
        instance,
        rerender,
      }
    })

    const rowModel = result.current.instance.getRowModel()

    expect(rowModel.rows.length).toEqual(3)
    expect(rowModel.flatRows.length).toEqual(3)
    expect(rowModel.rowsById['2']?.original).toEqual(defaultData[2])
  })
})
