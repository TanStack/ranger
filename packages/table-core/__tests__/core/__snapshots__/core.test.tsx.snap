// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`core renders a table with markup 1`] = `
[
  [
    [
      "Name",
      "2",
    ],
    [
      "Info",
      "4",
    ],
  ],
  [
    [
      "",
      "1",
    ],
    [
      "",
      "1",
    ],
    [
      "",
      "1",
    ],
    [
      "More Info",
      "3",
    ],
  ],
  [
    [
      "firstName",
      "1",
    ],
    [
      "<span>Last Name</span>",
      "1",
    ],
    [
      "Age",
      "1",
    ],
    [
      "<span>Visits</span>",
      "1",
    ],
    [
      "Status",
      "1",
    ],
    [
      "Profile Progress",
      "1",
    ],
  ],
]
`;

exports[`core renders a table with markup 2`] = `
[
  [
    "tanner",
    "linsley",
    "29",
    "100",
    "In Relationship",
    "50",
  ],
  [
    "derek",
    "perkins",
    "40",
    "40",
    "Single",
    "80",
  ],
  [
    "joe",
    "bergevin",
    "45",
    "20",
    "Complicated",
    "10",
  ],
]
`;

exports[`core renders a table with markup 3`] = `
[
  [
    [
      "firstName",
      "1",
    ],
    [
      "lastName",
      "1",
    ],
    [
      "age",
      "1",
    ],
    [
      "visits",
      "1",
    ],
    [
      "status",
      "1",
    ],
    [
      "progress",
      "1",
    ],
  ],
  [
    [
      "",
      "1",
    ],
    [
      "",
      "1",
    ],
    [
      "",
      "1",
    ],
    [
      "More Info",
      "3",
    ],
  ],
  [
    [
      "Name",
      "2",
    ],
    [
      "Info",
      "4",
    ],
  ],
]
`;
